apache2/waf_logging/waf_log_util.cc

```diff
@@ -0,0 +1,131 @@
+
+// This function fills in a waf format message based on modsec input.
+void set_waf_format(waf_format::Waf_Format* waf_format, char* clientIP, char* clientPort, char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, char* hostname, char* time) {
+    if (clientIP != NULL) {
+        waf_format->set_clientip(clientIP);
+    }
+    
+    if (clientPort != NULL) {
+        waf_format->set_clientport(clientPort);
+    }
+    
+    if (requestUri != NULL) {
+        waf_format->set_requesturi(requestUri);
+    }
+    
+    if (ruleSetType != NULL) {
+        waf_format->set_rulesettype(ruleSetType);
+    }
+    
+    if (ruleSetVersion != NULL) {
+        waf_format->set_rulesetversion(ruleSetVersion);
+    }
+    
+    if (ruleId != NULL) {
+        waf_format->set_ruleid(ruleId);
+    }
+    
+    if (messages != NULL) {
+        waf_format->set_messages(messages);
+    }
+    
+    switch(action) {
+        case 1:
+            waf_format->set_action(waf_format::Waf_Format::Detected);
+            break;
+        case 2:
+            waf_format->set_action(waf_format::Waf_Format::Blocked);
+            break;
+        default:
+            break;
+    }
+    
+    if (site == 0) {
+        waf_format->set_site(waf_format::Waf_Format::Global);
+    }
+    
+    waf_format::Details details;
+    
+    if (details_messages != NULL) {
+        details.set_messages(details_messages);
+    }
+    
+    if (details_data != NULL) {
+        details.set_data(details_data);
+    }
+    
+    if (details_file != NULL) {
+        details.set_file(details_file);
+    }
+    
+    if (details_line != NULL) {
+        details.set_line(details_line);
+    }
+    
+    *waf_format->mutable_details() = details;
+    
+    if (hostname != NULL) {
+        waf_format->set_hostname(hostname);
+    }
+
+    if (time != NULL) {
+        waf_format->set_time(time);
+    }
+}
+
+// Main function:  get fields from modsec, set the protobuf object and write to file in json.
+int write_json_to_file(char* data_dir, char* clientIP, char* clientPort, char* requestUri, char* ruleSetType, char* ruleSetVersion, char* ruleId, char* messages, int action, int site, char* details_messages, char* details_data, char* details_file, char* details_line, char* hostname, char* time) {
+    waf_format::Waf_Format waf_format;
+    std::ofstream json_file;
+    std::string json_string;
+    google::protobuf::util::JsonPrintOptions options;
+    google::protobuf::util::Status convert_result;
+    std::string directory_name;
+    std::string file_name;
+    const char* fullname;
+    
+    if (data_dir == NULL) {
+       return WAF_LOG_UTIL_FAILED;
+    }
+    
+    directory_name = data_dir;
+    file_name = WAF_LOG_UTIL_FILE;
+    fullname = (directory_name + "/" + file_name).c_str();
+
+    json_file.open(fullname, std::ios::app);
```

Comment:
should we perhaps open this only once, and not on every single write?

Reply:
yes, made the change